safehttp/task.go

```diff
@@ -0,0 +1,256 @@
+	// respond with a 500 Internal Server Error.
+	defer func() {
+		if r := recover(); r != nil {
+			t.WriteError(StatusInternalServerError)
+		}
+	}()
+
+	for _, it := range t.cfg.Interceptors {
+		it.Before(t, t.req)
+		if t.written {
+			return
+		}
+	}
+
+	t.cfg.Handler.ServeHTTP(t, t.req)
+	if !t.written {
+		t.NoContent()
+	}
+}
+
+// Write dispatches the response to the Dispatcher. This will be written to the
+// underlying http.ResponseWriter if the Dispatcher decides it's safe to do so.
+//
+// TODO: replace panics with proper error handling when writing the response fails.
+func (t *Task) Write(resp Response) Result {
+	if t.written {
+		panic("ResponseWriter was already written to")
+	}
+	t.written = true
+	t.commitPhase(resp)
+
+	ct, err := t.cfg.Dispatcher.ContentType(resp)
+	if err != nil {
+		panic(err)
+	}
+	t.rw.Header().Set("Content-Type", ct)
+
+	if t.code == 0 {
+		t.code = StatusOK
+	}
+	t.rw.WriteHeader(int(t.code))
+
+	if err := t.cfg.Dispatcher.Write(t.rw, resp); err != nil {
+		panic(err)
+	}
+	return Result{}
+}
+
+// NoContent responds with a 204 No Content response.
+//
+// If the ResponseWriter has already been written to, then this method will panic.
+func (t *Task) NoContent() Result {
+	if t.written {
+		panic("ResponseWriter was already written to")
+	}
+	t.written = true
+	t.commitPhase(NoContentResponse{})
+	t.rw.WriteHeader(int(StatusNoContent))
+	return Result{}
+}
+
+// WriteError writes an error response (400-599) according to the provided
+// status code.
+//
+// If the ResponseWriter has already been written to, then this method will panic.
+func (t *Task) WriteError(code StatusCode) Result {
+	// TODO: accept custom error responses that need to go through the dispatcher.
+	if t.writtenError {
+		panic("ResponseWriter.WriteError called twice")
+	}
+	t.written = true
+	t.writtenError = true
+	resp := &ErrorResponse{Code: code}
+	t.errorPhase(resp)
+	http.Error(t.rw, http.StatusText(int(resp.Code)), int(resp.Code))
+	return Result{}
+}
+
+// Redirect responds with a redirect to the given url, using code as the status code.
+//
+// If the ResponseWriter has already been written to, then this method will panic.
+func (t *Task) Redirect(r *IncomingRequest, url string, code StatusCode) Result {
+	if code < 300 || code >= 400 {
+		panic("wrong method called")
+	}
+	if t.written {
+		panic("ResponseWriter was already written to")
+	}
+	t.written = true
+	http.Redirect(t.rw, r.req, url, int(code))
+	return Result{}
+}
+
+// Header returns the collection of headers that will be set
+// on the response. Headers must be set before writing a
+// response (e.g. Write, WriteTemplate).
+func (t *Task) Header() Header {
+	return t.header
+}
+
+// SetCookie adds a Set-Cookie header to the provided ResponseWriter's headers.
+// The provided cookie must have a valid Name, otherwise an error will be
+// returned.
+func (t *Task) SetCookie(c *Cookie) error {
```

Comment:
Maybe we should change this to `AddCookie`?

Reply:
Probably, I'll leave a TODO.